rbac/management/group/view.py

```diff
@@ -669,19 +677,41 @@ def principals(self, request: Request, uuid: Optional[UUID] = None):
 
             # Process user principals and add them to the group.
             if len(principals) > 0:
-                resp = self.add_principals(group, principals, org_id=org_id)
+                proxy_response = self.validate_principals_in_proxy_request(principals, org_id=org_id)
 
-            # Storing user principals might return an error structure instead of a group,
-            # so we need to check that before returning a response.
-            if isinstance(resp, dict) and "errors" in resp:
-                return Response(status=resp["status_code"], data=resp["errors"])
+            with transaction.atomic():
+                group = self._get_locked_object()
```

Comment:
I'm actually not sure we need to lock the group?

The main thing we're worried about is the outbox being out of order.

In the role case, the outbox event was _dependent_ upon all events before it, because it referenced existing IDs that could have been changed by the very previous / concurrent tx.

In the group principal case, the IDs are always the same – group, user & service account ID – regardless of previous state. So I think we should be able to get away without locking the group too in this case.

For example, there can definitely be race conditions but I don't think they "matter" in that, though they may appear odd in some cases, they still result in consistent relations. For example:

1. tx1 add user1 start
2. tx2 add user1 start
3. tx1 add user commit – outbox add user1
4. tx3 remove user start
5. tx3 remove user commit – outbox remove user1
6. tx2 add user commit – outbox add user1 

This would result in a user added even though we just removed them, but the rbac db & relations would still be consistent about this. It's the same outcome as if tx2 just started a "little" bit later and was after tx3 instead of interleaving.

Locking would change the order (tx2 would block until tx1 committed, tx3 would block until tx2 committed, and therefore we'd end up with the user removed instead of added), but again this is only the difference in slight timings. It doesn't change the consistency of the data.

The main thing I think is just that any group changes MUST be committed with the associated outbox event wholly representing those changes in the same transaction.

Can you think of a case where we need to lock the group?

Reply:
Thanks @alechenninger for explaining! 

Mainly I was thinking if group is deleted during current transaction - do you think it is reason to lock group ?

1. tx1 add user1 start
2. tx2 delete group
3. tx1 sent data to outbox table with group which doesn't exist.

This was my reasoning

Reply:
Yeah great question. The answer to this is basically the same for the principal case. I think you're talking about this scenario:

Tx1 | Tx2 | Result | Input | Output | Tuple state
-- | -- | -- | -- | -- | --
check if g1 exists |   | ok | g1 | exists |  
add user1 g1 |   | ok | u1, g1 |   |  
add outbox event |   | ok |   |   |  
  | check if g1 exists | ok |   | exists |  
  | remove g1 | ok | g1 |   |  
  | select for update role bindings w/ g1 | ok | g1 | role binding list |  
  | update / remove role bindings | ok | new role binding state |   |  
  | add outbox event | ok |   |   |  
  | commit | ok |   |   | ok
commit |   | ok |   |   | not ok! outbox event adds orphaned group#member@user tuple

I don't think that is actually what would happen. The reason is because of foreign key constraints. 

```
ALTER TABLE "management_group_principals" ADD CONSTRAINT "management_group_pri_group_id_0c1b3b8d_fk_managemen" FOREIGN KEY ("group_id") REFERENCES "management_group" ("id") DEFERRABLE INITIALLY DEFERRED; (params None)

ALTER TABLE "management_group_principals" ADD CONSTRAINT "management_group_pri_principal_id_47b7a6f8_fk_managemen" FOREIGN KEY ("principal_id") REFERENCES "management_principal" ("id") DEFERRABLE INITIALLY DEFERRED; (params None)
```

What should happen is this:

Tx1 | Tx2 | Result | Input | Output | Tuple state
-- | -- | -- | -- | -- | --
check if g1 exists |   | ok | g1 | exists |  
add user1 g1 |   | ok | u1, g1 |   |  
add outbox event |   | ok |   |   |  
  | check if g1 exists | ok |   | exists |  
  | remove g1 | ok | g1 |   |  
  | select for update role bindings w/ g1 | ok | g1 | role binding list |  
  | update / remove role bindings | ok | new role binding state |   |  
  | add outbox event | ok |   |   |  
  | commit | ok |   |   | ok
commit |   | rollback - fk constraint violation |   |   | ok – no event committed (rollback)

When tx1 goes to commit, it shouldn't succeed because of the foreign key constraints b/w the group and the principal. Might be worth confirming this by manually running sql against two concurrent sessions in the db but I think that's what should happen. In that case, as long as the outbox event and the user add are in the same tx (atomic block), the outbox event actually never gets written because the user addition never gets written, either, all without locking.

That said, we can consider locking the group _anyway_ since it's not too expensive and might save us in case there is a scenario we aren't thinking of. But I think we're okay.

For principals, it's the same thing. If a concurrent tx removes the principal and commits, and then we try to refer to that principal with another commit, that will rollback b/c the principal no longer exists.

Reply:
Thanks ! I will leave here locking and principals here in that case.

Reply:
Not sure I follow – what I am describing is why the lock is not necessary. The first table isn't actually what would happen, I was clarifying the difference between I think what you were expecting vs what I think would actually happen.

It's not a huge deal to leave the locks but shouldn't be necessary. Are you saying you want to leave it just in case?

Reply:
yes I understand what did you describe but according to statement and I also confirmed it with 2 sessions.
in TX1, it tries to insert a group ID that no longer exists because it was deleted by transaction TX2, resulting in an error due to a constraint violation

I intended to leave it here because of this sentence:

> That said, we can consider locking the group anyway since it's not too expensive and might save us in case there is a scenario we aren't thinking of. But I think we're okay.

Now I understand that you prefer not to use locking for groups and principals, 
so I will remove it. 

NOTE: With locks, the only difference is that TX2 will wait for TX1, but most likely it will fail because TX2 is attempting to delete a group that now has a reference in the management_group_principals table, which was added by TX1.

Reply:
Thanks for clarifying and confirming locally!

I guess one another small downside to locking is that if there are other transactions trying to write to the group in some way, those will block while the group is locked (even if they are unrelated to dual write, e.g. group name or something like that). A small thing, though!